### PR TITLE
(CDAP-2772) (CDAP-5407) Cleanup the application deployment pipeline

### PR DESCRIPTION
**Pull request by chtyim**
- Get rid of copying of artifact jars to program jars in the program generation phase
- Cleanup the DefaultProgram to make the logic straight forward
- Make Program creation always using the artifacty jar the the application is created with
- Cleanup WorkflowProgramRunner logic
- Refactor DistributedProgramRunner and related classes logic
  - Serialize the program ID and the app spec so that it can recreate the program correctly

### CODE REVIEW

**Comment by chtyim** [comment]
Build already passed: http://builds.cask.co/browse/CDAP-DUT4138
Also test on cluster with purchase app and page rank app.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java [hunk 1]
```diff
@@ -55,35 +55,35 @@
...
-    this.name = app.getClass().getSimpleName();
-    this.description = "";
+  public DefaultAppConfigurer(Id.Namespace namespace, Id.Artifact artifactId, Application app) {
+    this(namespace, artifactId, app, "", null, null);
   }
 
   public DefaultAppConfigurer(Id.Namespace namespace, Id.Artifact artifactId, Application app, String configuration,
-                              ArtifactRepository artifactRepository, PluginInstantiator pluginInstantiator) {
+                              @Nullable ArtifactRepository artifactRepository,
+                              @Nullable  PluginInstantiator pluginInstantiator) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java]: nit: remove extra space

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ProgramDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        return (T) getApplicationSpecification().getWorkflows().get(id.getProgram());
+      case SERVICE:
+        return (T) getApplicationSpecification().getServices().get(id.getProgram());
+      case SPARK:
+        return (T) getApplicationSpecification().getSpark().get(id.getProgram());
+      case WORKER:
+        return (T) getApplicationSpecification().getWorkers().get(id.getProgram());
+      default:
+        // This shouldn't happen
+        throw new IllegalStateException("Unsupport program type " + id.getType());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ProgramDescriptor.java]: nit: Unsupport -> Unsupported

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java [hunk 3]
```diff
@@ -20,58 +20,35 @@
...
-   * @param unpackedDir a directory that the program jar file was unpacked to
-   * @return a new {@link Program} instance.
-   * @throws IOException If failed to create the program
-   */
-  public static Program create(CConfiguration cConf, ProgramRunner programRunner,
+    * Creates a {@link Program} that can be executed by the given {@link ProgramRunner}.
+    *
+    * @param cConf the CDAP configuration
+    * @param programRunner the {@link ProgramRunner} for executing the program. If provided and if it implements
+   *                       {@link ProgramClassLoaderProvider}, then the
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java]: nit: fix indentation, '*' is off by a space

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java [hunk 4]
```diff
@@ -20,58 +20,35 @@
...
+   *                       {@link ClassLoader} created for the {@link Program} will be determined based on it.
+   *                       Otherwise, the {@link ClassLoader} will only have visibility
+   *                       to cdap-api and hadoop classes.
+   *  @param programDescriptor description of the program to create
+    * @param programJarLocation the {@link Location} of the program jar file
+    * @param unpackedDir a directory that the program jar file was unpacked to
+    * @return a new {@link Program} instance.
+   -   * @throws IOException If failed to create the program
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java]: remove '-'

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java [hunk 5]
```diff
@@ -207,9 +217,9 @@ private ProgramOptions createPluginSnapshot(ProgramOptions options, Id.Program p
       }
 
       try {
-        ArtifactDetail detail = artifactRepository.getArtifact(Id.Artifact.from(programId.getNamespace(),
-                                                                                plugin.getArtifactId()));
-        Files.copy(Locations.newInputSupplier(detail.getDescriptor().getLocation()), destFile);
+        ArtifactId artifactId = Artifacts.toArtifactId(programId.getNamespaceId(), plugin.getArtifactId());
+        ArtifactDetail detail = artifactRepository.getArtifact(artifactId.toId());
+        Locations.linkOrCopy(detail.getDescriptor().getLocation(), destFile);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java:222]: linking has a larger window for a race where a snapshot artifact is re-deployed after this happens, then the original is deleted. Though I guess this only happens in the sdk?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java:222]: This is a hardlink being created, how can it has a larger window for race??

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DatasetModulesDeployer.java [hunk 6]
```diff
@@ -38,32 +40,31 @@
...
   private final DatasetFramework datasetFramework;
   // An instance of InMemoryDatasetFramework is used to check if a dataset is a system dataset
   private final DatasetFramework systemDatasetFramework;
-  private final Id.Namespace namespace;
   private final boolean allowDatasetUncheckedUpgrade;
 
-  public DatasetModulesDeployer(DatasetFramework datasetFramework, DatasetFramework inMemoryDatasetFramework,
-                                Id.Namespace namespace, CConfiguration cConf) {
+  DatasetModulesDeployer(DatasetFramework datasetFramework, DatasetFramework
+    inMemoryDatasetFramework, CConfiguration cConf) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DatasetModulesDeployer.java]: nit: this is a weird line break, inMemoryDatasetFramework can be on the previous line or entirely on this line

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ProgramGenerationStage.java [hunk 7]
```diff
@@ -100,125 +82,13 @@ public void process(final ApplicationDeployable input) throws Exception {
       appSpec.getWorkers().values()
     );
 
-    // Generate webapp program if required
-    Set<String> servingHostNames = WebappProgramRunner.getServingHostNames(
-      Locations.newInputSupplier(input.getLocation()));
-
-    if (!servingHostNames.isEmpty()) {
-      specifications = Iterables.concat(specifications, ImmutableList.of(
-        createWebappSpec(ProgramType.WEBAPP.toString().toLowerCase())));
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ProgramGenerationStage.java:109]: should we remove WEBAPP from ProgramType then?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java (hunk 5 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java:222]: oh just assumed it was a symlink, ignore

**Comment by albertshau** [comment]
some minor comments, lgtm

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ProgramDescriptor.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/ProgramDescriptor.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java (hunk 3 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java (hunk 4 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/program/Programs.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DatasetModulesDeployer.java (hunk 6 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DatasetModulesDeployer.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ProgramGenerationStage.java (hunk 7 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ProgramGenerationStage.java:109]: Maybe. Just afraid after removing it, there will be some deserialization error.

**Comment by chtyim** [comment]
Addressed comments. All comments involve only spacing and javadoc. Since test already passed, squash and merge.

**Pull request merged**